### PR TITLE
Adding batch_update_findings functionality to AWS SecurityHub integration

### PR DESCRIPTION
**Pull request by danielperez660**
## 🧩 Type of Change

- [x] Feature
- [ ] Bug fix
- [ ] Other

## 🧠 Background & Motivation
We can only read from Security Hub whereas to fully integrate with it, we should also be able to write to it. To do this, I have integrated the boto3 functionality of `batch_update_findings` - which will allow users to write and close findings directly from Rapid7 rather than having to do this manually or rolling out their own keystore/auth in python connections, which is the current requirement.

## ✨ What Changed
Added a new action to the aws_securityhub plugin to allow updating of securityhub findings.

## 🧪 Testing
Testing has been done manually to confirm that the findings IDs which are passed to the Rapid7 connection are updated where needed
image
image
image

### CODE REVIEW

**Comment by danielperez660** [comment]
Hey @ewilson-r7 are you, or anyone in your team free to look into this? 
Thank you very much!

File: plugins/aws_securityhub/icon_aws_securityhub/actions/batch_update_findings/action.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            "VerificationState": verification_state,
+            "Confidence": confidence,
+            "Criticality": criticality,
+            "Types": types,
+            "UserDefinedFields": user_defined_fields,
+            "Workflow": workflow,
+            "RelatedFindings": related_findings,
+        }
+
+        filtered_params = {k: v for k, v in update_params.items() if v}
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/icon_aws_securityhub/actions/batch_update_findings/action.py]: 🔴 **Critical: Falsy filter drops valid values + single-letter variable names**

Two issues:
1. `if v` silently drops legitimate values like `Confidence: 0`, `Criticality: 0`, empty lists `[]`, etc. Zero is a valid confidence value per the Security Hub API.
2. Single-character variable names (`k`, `v`) don't meet our naming standards — use descriptive identifiers even in comprehensions.

```suggestion
        filtered_params = {param: value for param, value in update_params.items() if value is not None}
```

File: plugins/aws_securityhub/icon_aws_securityhub/actions/batch_update_findings/action.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                unprocessed_findings.append(finding)
+
+        if results.get("ProcessedFindings"):
+            for finding in results["ProcessedFindings"]:
+                processed_findings.append(finding)
+
+        return {
+            Output.PROCESSED_FINDINGS: processed_findings,
+            Output.UNPROCESSED_FINDINGS: unprocessed_findings,
+        }
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/icon_aws_securityhub/actions/batch_update_findings/action.py:68]: 🔴 **Critical: No error handling**

The `run()` method has zero exception handling. Per our standards, all API errors must be wrapped in `PluginException` — never let raw exceptions propagate. Also, the list-building loops (lines 58-64) are unnecessarily verbose.

Suggested rewrite of the full `run()` method:

```suggestion
    def run(self, params={}):
        from insightconnect_plugin_runtime.exceptions import PluginException
        from botocore.exceptions import ClientError

        finding_identifiers = params.get(Input.FINDING_IDENTIFIERS)
        note = params.get(Input.NOTE, None)
        severity = params.get(Input.SEVERITY, None)
        verification_state = params.get(Input.VERIFICATION_STATE, None)
        confidence = params.get(Input.CONFIDENCE, None)
        criticality = params.get(Input.CRITICALITY, None)
        types = params.get(Input.TYPES, None)
        user_defined_fields = params.get(Input.USER_DEFINED_FIELDS, None)
        workflow = params.get(Input.WORKFLOW, None)
        related_findings = params.get(Input.RELATED_FINDINGS, None)

        update_params = {
            "FindingIdentifiers": finding_identifiers,
            "Note": note,
            "Severity": severity,
            "VerificationState": verification_state,
            "Confidence": confidence,
            "Criticality": criticality,
            "Types": types,
            "UserDefinedFields": user_defined_fields,
            "Workflow": workflow,
            "RelatedFindings": related_findings,
        }

        filtered_params = {param: value for param, value in update_params.items() if value is not None}
        client = self.connection.aws.client("securityhub")

        try:
            results = client.batch_update_findings(**filtered_params)
        except ClientError as error:
            raise PluginException(
                cause="AWS Security Hub API call failed.",
                assistance=f"Verify your credentials and finding identifiers. Error: {error.response['Error']['Message']}",
                data=error,
            )

        return {
            Output.PROCESSED_FINDINGS: results.get("ProcessedFindings", []),
            Output.UNPROCESSED_FINDINGS: results.get("UnprocessedFindings", []),
        }
```

File: plugins/aws_securityhub/icon_aws_securityhub/actions/batch_update_findings/action.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+from .schema import (
+    BatchUpdateFindingsInput,
+    BatchUpdateFindingsOutput,
+    Input,
+    Output,
+    Component,
+)
+
+# Custom imports below
+import logging
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/icon_aws_securityhub/actions/batch_update_findings/action.py:14]: 🟢 **Minor:** Module-level `logging.getLogger("botocore").setLevel(logging.CRITICAL)` mutates global logging state at import time. I know it's copied from `get_findings`, but ideally this would be configured once in the connection's `connect()`. Low priority since it matches existing pattern.

File: plugins/aws_securityhub/unit_test/test_batch_update_findings.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+class TestBatchUpdateFindings(TestCase):
+    def test_batch_update_findings(self):
+        """
+        DO NOT USE PRODUCTION/SENSITIVE DATA FOR UNIT TESTS
+
+        TODO: Implement test cases here
+        """
+
+        self.fail("Unimplemented Test Case")
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/unit_test/test_batch_update_findings.py]: 🔴 **Critical: Placeholder test will fail the suite**

`self.fail("Unimplemented Test Case")` will actively break CI. We require minimum 80% coverage on new/modified code.

At minimum, mock the boto3 client and test:
- Successful batch update (happy path with processed findings)
- Handling of `UnprocessedFindings` in response
- Handling of `None`/missing optional params being filtered correctly
- Zero-value edge case (e.g., `Confidence: 0` must not be dropped)
- `ClientError` from boto3 is caught and wrapped in `PluginException`

Unused imports (`json`, `logging`, `Connection`) should also be removed — Prospector will flag them.

File: plugins/aws_securityhub/unit_test/test_get_findings.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+class TestGetFindings(TestCase):
+    def test_get_findings(self):
+        """
+        DO NOT USE PRODUCTION/SENSITIVE DATA FOR UNIT TESTS
+
+        TODO: Implement test cases here
+        """
+
+        self.fail("Unimplemented Test Case")
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/unit_test/test_get_findings.py]: 🟡 This is a new placeholder test file for an existing action — it will fail the suite as-is. If you didn't intend to add tests for `get_findings` in this PR, consider removing this file. If you do want to add them, they'll need actual implementation.

Same note about unused imports (`json`, `logging`).

File: plugins/aws_securityhub/plugin.spec.yaml [hunk 6]
```diff
@@ -1009,6 +1039,83 @@ actions:
...
+  batch_update_findings:
+    title: Batch Update Findings
+    description: Updates Security Hub-aggregated findings based on attributes
+    input:
+      finding_identifiers:
+        title: Finding Identifiers
+        description: An object of finding identifiers
+        type: "[]FindingIdentifiers"
+        required: true
+        example: [{"Id":arn:aws:securityhub:us-east-2:000000000000:subscription/cis-aws-foundations-benchmark/v/1.2.0/1.2/finding/0000000-0000-0000-0000-0000000000000", "ProductArn": "arn:aws:securityhub:us-east-2::product/aws/securityhub"}]
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/plugin.spec.yaml]: 🟡 **Invalid YAML — missing opening quote**

The value after `"Id":` is missing its opening quote mark. Same issue in the `related_findings` example further down. These will break `insight-plugin validate`.

```suggestion
        example: [{"Id": "arn:aws:securityhub:us-east-2:000000000000:subscription/cis-aws-foundations-benchmark/v/1.2.0/1.2/finding/0000000-0000-0000-0000-0000000000000", "ProductArn": "arn:aws:securityhub:us-east-2::product/aws/securityhub"}]
```

File: plugins/aws_securityhub/plugin.spec.yaml [hunk 7]
```diff
@@ -1009,6 +1039,83 @@ actions:
...
+      severity:
+          title: Severity
+          description: Used to update the finding severity
+          type: object
+          required: false
+          example: {"Normalized": 50, "Product": 8, "Label": "HIGH"}
+      verification_state:
+          title: Verification State
+          description: Indicates the veracity of a finding
+          type: VerificationState
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/plugin.spec.yaml]: 🟡 **Undefined type reference**

`VerificationState` is referenced as a type but never defined in the `types:` section. Since this is a constrained string per the AWS API, this should be:

```suggestion
        type: string
        enum:
          - UNKNOWN
          - TRUE_POSITIVE
          - FALSE_POSITIVE
          - BENIGN_POSITIVE
```

File: plugins/aws_securityhub/plugin.spec.yaml [hunk 8]
```diff
@@ -1009,6 +1039,83 @@ actions:
...
+          required: false
+          example: "FALSE_POSITIVE"
+      confidence:
+          title: Confidence
+          description: The updated value for the finding confidence
+          type: Confidence
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/plugin.spec.yaml]: 🟡 **Undefined type reference**

`Confidence` is not a custom type — per the AWS API this is simply an integer (0-100).

```suggestion
        type: integer
```

File: plugins/aws_securityhub/plugin.spec.yaml [hunk 9]
```diff
@@ -1009,6 +1039,83 @@ actions:
...
+          required: false
+          example: 100
+      criticality:
+          title: Criticality
+          description: The updated value for the level of importance assigned to the resources associated with the findings
+          type: Criticality
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/plugin.spec.yaml]: 🟡 Same issue — `Criticality` is an integer (0-100) per the AWS API, not a custom type.

```suggestion
        type: integer
```

File: plugins/aws_securityhub/plugin.spec.yaml [hunk 10]
```diff
@@ -1009,6 +1039,83 @@ actions:
...
+          required: false
+          example: 100
+      types:
+          title: Types
+          description: One or more finding types in the format of namespace/category/classifier that classify a finding
+          type: Types
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/plugin.spec.yaml]: 🟡 `Types` is not a custom type — this should be `"[]string"` per the API.

```suggestion
        type: "[]string"
```

File: plugins/aws_securityhub/plugin.spec.yaml [hunk 11]
```diff
@@ -1009,6 +1039,83 @@ actions:
...
+          required: false
+          example: ["TTPs/Command and Control/CryptoCurrency:EC2-BitcoinTool.B!DNS"]
+      user_defined_fields:
+          title: User Defined Fields
+          description: A list of name/value string pairs associated with the finding. These are custom, user-defined fields added to a finding
+          type: UserDefinedFields
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/plugin.spec.yaml]: 🟡 `UserDefinedFields` is not defined as a custom type. This is a string-to-string map, so `object` is acceptable here, but update the type reference:

```suggestion
        type: object
```

File: plugins/aws_securityhub/plugin.spec.yaml [hunk 12]
```diff
@@ -1009,6 +1039,83 @@ actions:
...
+          required: false
+          example: {"key": "value"}
+      workflow:
+          title: Workflow
+          description: Used to update the workflow status of a finding
+          type: WorkflowState
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/plugin.spec.yaml]: 🟡 `WorkflowState` is not a defined custom type. This should be a proper custom type with a `Status` field, or inline as `object`. Consider defining a `Workflow` type in the types section with:
```yaml
Workflow:
  Status:
    title: Status
    type: string
    description: Workflow status
    required: true
    enum:
      - NEW
      - NOTIFIED
      - RESOLVED
      - SUPPRESSED
```

Then reference it as `type: Workflow` here.

File: plugins/aws_securityhub/plugin.spec.yaml [hunk 13]
```diff
@@ -1009,6 +1039,83 @@ actions:
...
+          required: false
+          example: {"Status": "NEW"}
+      related_findings:
+          title: Related Findings
+          description: A list of findings that are related to the updated findings
+          type: object
+          required: false
+          example: {"Id":arn:aws:securityhub:us-east-2:000000000000:subscription/cis-aws-foundations-benchmark/v/1.2.0/1.2/finding/0000000-0000-0000-0000-0000000000000", "ProductArn": "arn:aws:securityhub:us-east-2::product/aws/securityhub"}
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/plugin.spec.yaml]: 🟡 Same YAML quoting issue as `finding_identifiers` — missing opening quote after `"Id":`. Also, `related_findings` should use the existing `[]RelatedFindings` type rather than raw `object`, since that type is already defined in this spec.

File: plugins/aws_securityhub/plugin.spec.yaml [hunk 14]
```diff
@@ -379,6 +404,11 @@ types:
       type: integer
       description: Product
       required: false
+    Label:
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/plugin.spec.yaml:418]: 🟡 **Copy-paste error — wrong title**

The `Label` field has `title: Product` but should be `title: Label`.

```suggestion
    Label:
      title: Label
      type: string
      description: Label
      required: false
```

File: plugins/aws_securityhub/icon_aws_securityhub/actions/__init__.py [hunk 15]
```diff
@@ -1,3 +1,5 @@
 # GENERATED BY INSIGHT-PLUGIN - DO NOT EDIT
 
 from .get_findings.action import GetFindings
+
+from .batch_update_findings.action import BatchUpdateFindings
```
Comment (ewilson-r7) [inline plugins/aws_securityhub/icon_aws_securityhub/actions/__init__.py:5]: 🟢 This is a generated file (`# GENERATED BY INSIGHT-PLUGIN - DO NOT EDIT`). The extra blank line and leading space are formatting artifacts from hand-editing. Run `insight-plugin refresh` to regenerate this cleanly.

**Review by ewilson-r7 — CHANGES_REQUESTED** [review]
Hey @danielperez660, thanks for putting this together — the feature fills a real gap and the approach is solid. I've left inline comments on the specific areas that need attention before we can merge.

## Summary

| Category | Count |
|----------|-------|
| 🔴 Critical (must fix) | 3 |
| 🟡 Moderate (should fix) | 5 |
| 🟢 Minor (nice to fix) | 3 |

Biggest concerns: the falsy-value filter bug, missing error handling, empty unit tests, and several spec YAML issues.

File: plugins/aws_securityhub/icon_aws_securityhub/actions/batch_update_findings/action.py (hunk 3 above)
Comment (danielperez660) [reply plugins/aws_securityhub/icon_aws_securityhub/actions/batch_update_findings/action.py:14]: Done

File: plugins/aws_securityhub/plugin.spec.yaml (hunk 11 above)
Comment (danielperez660) [reply plugins/aws_securityhub/plugin.spec.yaml]: resolved

File: plugins/aws_securityhub/plugin.spec.yaml (hunk 13 above)
Comment (danielperez660) [reply plugins/aws_securityhub/plugin.spec.yaml]: Good catch on the quote missing, I didn't notice. 
This has now been fixed along with the object type.

File: plugins/aws_securityhub/plugin.spec.yaml (hunk 12 above)
Comment (danielperez660) [reply plugins/aws_securityhub/plugin.spec.yaml]: done, got it mixed up with a subfield of another type

**Comment by danielperez660** [comment]
@ewilson-r7 I think that's all of the issues addressed
